### PR TITLE
materialgram: init at 5.4.1.1

### PR DESCRIPTION
**Pull request by oluceps**
## Description of changes
Add https://github.com/kukuruzka165/materialgram

image

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by Aleksanaa** [comment]
You are faster than me 😄

File: pkgs/by-name/ma/materialgram/package.nix [hunk 1]
```diff
@@ -0,0 +1,46 @@
+{
+  tdesktop,
```
Comment (Guanran928) [inline pkgs/by-name/ma/materialgram/package.nix:2]: I remember aliases is not allowed in nixpkgs.

https://github.com/NixOS/nixpkgs/blob/cafba5f9aa857651d6e59ebec7f759e6c56dc107/pkgs/top-level/aliases.nix#L1309

**Comment by Aleksanaa** [comment]
> Would you like to be a co-maintainer? @Aleksanaa

Okay but I may not update it often. Durov told me my telegram account is very precious and fragile 😅

**Comment by Aleksanaa** [comment]
Sorry I forgot this. I've lost my telegram account. @NickCao could you help with this?

File: pkgs/by-name/ma/materialgram/package.nix [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+    src = fetchFromGitHub {
+      owner = "kukuruzka165";
+      repo = "materialgram";
+      rev = "v${finalAttrs.version}";
+      fetchSubmodules = true;
+      hash = "sha256-OlXcDsjZ59/GjysfFAOo6dAoNnUY5hBD9NuNgTR4rXc=";
+    };
+
+    postFixup =
```
Comment (NickCao) [inline pkgs/by-name/ma/materialgram/package.nix:21]: We shall not duplicate this part in every tdesktop fork.... Lemme do a quick preparatory PR to clean things up.

Comment (NickCao) [reply pkgs/by-name/ma/materialgram/package.nix:21]: https://github.com/NixOS/nixpkgs/pull/324630

**Comment by incycledream** [comment]
@Aleksanaa, Are you going to merge or not? Almost two months passed

**Comment by ilsubyeega** [comment]
@incycledream 
> Are you going to merge or not? Almost two months passed

it is currently broken with latest nixpkgs master branch, maybe locked with #331396 afterwards

**Comment by NickCao** [comment]
While at it could fellow telegram users test https://github.com/NixOS/nixpkgs/pull/331396 out? Especially around voice/video calling since I've reworked tg_owt a bit.

**Comment by incycledream** [comment]
Knock-knock, where's the merge? Telegram Desktop got already merged, I can't wait for materialgram in Nix to ditch Flatpak (Even considering that fact I was the person who got materialgram into Flathub). By the way, I can also be a co-maintainer of this package if required

**Comment by Aleksanaa** [comment]
@incycledream could you help me to test it? I'm a little afraid of my telegram account disappearing again. Just `nixpkgs-review pr 321531` and run the executable.

**Comment by incycledream** [comment]
@Aleksanaa 
> @incycledream could you help me to test it? I'm a little afraid of my telegram account disappearing again. Just `nixpkgs-review pr 321531` and run the executable.

I got an error while building
```
warning: The interpretation of store paths arguments ending in `.drv` recently changed. If this command is now failing try again with '/nix/store/dnc2xnz6d2279iiaalx9bsysd1hwb2dd-materialgram-5.2.1.1.drv^*'
Running phase: qtPreHook
@nix { "action": "setPhase", "phase": "qtPreHook" }
Running phase: unpackPhase
@nix { "action": "setPhase", "phase": "unpackPhase" }
unpacking source archive /nix/store/kf7na7f3lymipzbirhnpyf4s9dbqj7kv-source
source root is source
Running phase: patchPhase
@nix { "action": "setPhase", "phase": "patchPhase" }
applying patch /nix/store/3irbx4ih9kcwwfdpsv9i6n96dn0pa1q2-macos.patch
patching file Telegram/SourceFiles/platform/mac/overlay_widget_mac.mm
patching file Telegram/lib_base/base/platform/mac/base_battery_saving_mac.mm
patching file Telegram/lib_webrtc/webrtc/platform/mac/webrtc_environment_mac.mm
patching file Telegram/lib_webview/webview/platform/mac/webview_mac.mm
Hunk #1 succeeded at 342 with fuzz 2 (offset 28 lines).
Hunk #2 succeeded at 702 (offset 42 lines).
applying patch /nix/store/8paj2qx3abvp2mpd1y0llbknqir6ssvf-scheme.patch
patching file Telegram/SourceFiles/core/application.cpp
Hunk #1 FAILED at 421.
1 out of 1 hunk FAILED -- saving rejects to file Telegram/SourceFiles/core/application.cpp.rej
```
Here's the rej file:
```
--- Telegram/SourceFiles/core/application.cpp
+++ Telegram/SourceFiles/core/application.cpp
@@ -421,7 +421,6 @@ void Application::run() {
 
 void Application::autoRegisterUrlScheme() {
 	if (!OptionSkipUrlSchemeRegister.value()) {
-		InvokeQueued(this, [] { RegisterUrlScheme(); });
 	}
 }
 
```
Here's the application.cpp:
https://snips.sh/f/j4SnXH2nVY?r=1

**Comment by dvtfl** [comment]
Even with removed patch, building fails with 300 000 lines of log.
```
/nix/store/lbk30k56awz9vz9qpid93fkjns0xwlhd-gcc-wrapper-13.3.0/bin/g++ -DBWE_TEST_LOGGING_COMPILE_TIME_ENABLE=0 -DHAVE_WEBRTC_VIDEO -DNDEBUG -DNO_MAIN_THREAD_WRAPPING -DQT_DEPRECATED_WARNINGS_SINCE=0x051500 -DQT_IMPLICIT_QCHAR_CONSTRUCTION -DQT_NO_CAST_FROM_BYTEARRAY -DQT_NO_KEYWORDS -DRTC_DISABLE_TRACE_EVENTS -DRTC_ENABLE_H265 -DRTC_ENABLE_VP9 -DTGCALLS_USE_STD_OPTIONAL -DWEBRTC_APM_DEBUG_DUMP=0 -DWEBRTC_APP_TDESKTOP -DWEBRTC_DUMMY_AUDIO_BUILD -DWEBRTC_ENABLE_PROTOBUF=0 -DWEBRTC_HAVE_DCSCTP -DWEBRTC_HAVE_SCTP -DWEBRTC_INCLUDE_INTERNAL_AUDIO_DEVICE -DWEBRTC_LIBRARY_IMPL -DWEBRTC_LINUX -DWEBRTC_NON_STATIC_TRACE_EVENT_HANDLERS=1 -DWEBRTC_OPUS_SUPPORT_120MS_PTIME=1 -DWEBRTC_OPUS_VARIABLE_COMPLEXITY=0 -DWEBRTC_POSIX -DWEBRTC_USE_BUILTIN_ISAC_FLOAT -DWEBRTC_USE_H264 -DWEBRTC_USE_PIPEWIRE -DWEBRTC_USE_X11 -I/build/source/Telegram/ThirdParty/tgcalls -I/build/source/Telegram/ThirdParty/tgcalls/tgcalls -isystem /nix/store/2bw54660ndcc79c6zjm5g98slkpks814-tg_owt-0-unstable-2024-08-02-dev/include/tg_owt -isystem /nix/store/2bw54660ndcc79c6zjm5g98slkpks814-tg_owt-0-unstable-2024-08-02-dev/include/tg_owt/third_party/libyuv/include -O3 -DNDEBUG -std=gnu++20 -Wno-deprecated-volatile -Wno-ambiguous-reversed-operator -Wno-deprecated-declarations -Wno-unqualified-std-cast-call -fPIC -fno-strict-aliasing -pipe -Wall -Wextra -Wno-unused-parameter -Wno-switch -Wno-maybe-uninitialized -Wno-missing-field-initializers -Wno-sign-compare -Wno-deprecated -MD -MT Telegram/CMakeFiles/lib_tgcalls.dir/ThirdParty/tgcalls/tgcalls/LogSinkImpl.cpp.o -MF Telegram/CMakeFiles/lib_tgcalls.dir/ThirdParty/tgcalls/tgcalls/LogSinkImpl.cpp.o.d -o Telegram/CMakeFiles/lib_tgcalls.dir/ThirdParty/tgcalls/tgcalls/LogSinkImpl.cpp.o -c /build/source/Telegram/ThirdParty/tgcalls/tgcalls/LogSinkImpl.cpp
In file included from /build/source/Telegram/ThirdParty/tgcalls/tgcalls/LogSinkImpl.cpp:3:
/build/source/Telegram/ThirdParty/tgcalls/tgcalls/Instance.h:17:7: error: conflicting declaration of template 'template<class T> struct rtc::scoped_refptr'
   17 | class scoped_refptr;
      |       ^~~~~~~~~~~~~
In file included from /nix/store/2bw54660ndcc79c6zjm5g98slkpks814-tg_owt-0-unstable-2024-08-02-dev/include/tg_owt/rtc_base/checks.h:59,
``` 

[log](https://files.catbox.moe/pwkyrd.log)

**Comment by incycledream** [comment]
@oluceps @Aleksanaa
Here you go, it works! I'm so happy that I've even made a selfie with materialgram on NixOS. I need to propose some changes though (thanks to @dvtfl and her friend),  but I'm gonna do that tomorrow
IMG_20240819_235316_1

**Comment by Kamillaova** [comment]
@oluceps @Aleksanaa [package.nix](https://gist.github.com/Kamillaova/3ca3bca61d3d7568cc5f6cc52a835267)

this version works, I think the only changed part is the version

tested on x86_64-linux (cc @dvtfl @incycledream), aarch64-darwin

image

File: pkgs/by-name/ma/materialgram/package.nix [hunk 3]
```diff
@@ -0,0 +1,34 @@
+{
+  telegram-desktop,
+  lib,
+  fetchFromGitHub,
+}:
+telegram-desktop.overrideAttrs (
+  finalAttrs: previousAttrs: {
+    pname = "materialgram";
+    version = "5.2.1.1";
```
Comment (Kamillaova) [inline pkgs/by-name/ma/materialgram/package.nix:9]:
```suggestion
    version = "5.4.1.1";
```

and also
```diff
       fetchSubmodules = true;
-      hash = "sha256-tofpm5oz4E7UaGw1rD39UF0c22sxhazLm9ZvKX0vHSk=";
+      hash = "sha256-enA/8mIXcRTPGyZFgZ5Wg4fWq4fD4rtjH7bm/32FZaE";
     };
```

**Review by Kamillaova — CHANGES_REQUESTED** [review]

**Comment by Kamillaova** [comment]
why does this pr has `11.by: package-maintainer` label?

**Comment by Kamillaova** [comment]
@oluceps can you update pr name?

**Comment by oluceps** [comment]
> why does this pr has `11.by: package-maintainer` label? does this 

That may bcz I'm a package maintainer

**Comment by Kamillaova** [comment]
https://github.com/NixOS/nixpkgs/pull/334732 whoops

**Review by NickCao — APPROVED** [review]
diff LGTM, please test!

**Review by pan93412 — APPROVED** [review]
CleanShot 2024-08-22 at 20 35 11@2x
CleanShot 2024-08-22 at 20 36 32@2x

Tested it on aarch64-darwin. Works.

**Review by Aleksanaa — APPROVED** [review]

**Comment by dvtfl** [comment]
Result of `nixpkgs-review pr 321531` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>materialgram</li>
  </ul>
</details>

**Review by s0me1newithhand7s — APPROVED** [review]
can't wait it in `nixpkgs`, eager to compare it with `_64gram`.

`nixpkgs-review pr 321531`:
1 package built:
materialgram

**Review by ViZiD — APPROVED** [review]

**Review by stellessia — APPROVED** [review]

**Review by kukuruzka165 — APPROVED** [review]

**Review by nuwawa — APPROVED** [review]
Audio, video and screen sharing in calls works on linux.

**Comment by Kamillaova** [comment]
I'll test voice+video+screenshare calls on aarch64-darwin

```
[546/1410] Building CXX object Telegram/CMakeFiles/td_ui.dir/SourceFiles/statistics/chart_widget.cpp.o
[547/1410] Building CXX object Telegram/CMakeFiles/td_ui.dir/SourceFiles/ui/boxes/auto_delete_settings.cpp.o
```

**Review by avapievna — APPROVED** [review]

**Review by ghost — APPROVED** [review]

**Review by Kamillaova — APPROVED** [review]
Voice, video, screen share calls works in fullduplex, tested on aarch64-darwin

image

**Comment by Guanran928** [comment]
Result of `nixpkgs-review pr 321531` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>materialgram</li>
  </ul>
</details>

**Review by Guanran928 — APPROVED** [review]

**Review by gmankab — APPROVED** [review]

**Pull request merged**